stdlib/public/core/StringComparison.swift

```diff
@@ -0,0 +1,1368 @@
+  }
+  let leftHSPtr = leftHS.baseAddress._unsafelyUnwrappedUnchecked
+  let rightHSPtr = rightHS.baseAddress._unsafelyUnwrappedUnchecked
+  return _lexicographicalCompare(leftHSPtr[idx], rightHSPtr[idx])
+}
+@inline(__always)
+internal func _lexicographicalCompare(
+  _ leftHS: UnsafeBufferPointer<UInt8>,
+  _ rightHS: UnsafeBufferPointer<UInt16>
+) -> _Ordering {
+  let count = Swift.min(leftHS.count, rightHS.count)
+
+  let idx = _findDiffIdx(leftHS, rightHS)
+  guard idx < count else {
+    return _lexicographicalCompare(leftHS.count, rightHS.count)
+  }
+  let leftHSPtr = leftHS.baseAddress._unsafelyUnwrappedUnchecked
+  let rightHSPtr = rightHS.baseAddress._unsafelyUnwrappedUnchecked
+  return _lexicographicalCompare(UInt16(leftHSPtr[idx]), rightHSPtr[idx])
+}
+@inline(__always)
+internal func _lexicographicalCompare(
+  _ leftHS: UnsafePointer<_Normalization._SegmentOutputBuffer>,
+  leftCount: Int,
+  _ rightHS: UnsafePointer<_Normalization._SegmentOutputBuffer>,
+  rightCount: Int
+) -> _Ordering {
+  return _lexicographicalCompare(
+    _castOutputBuffer(leftHS, endingAt: leftCount),
+    _castOutputBuffer(rightHS, endingAt: rightCount))
+}
+@inline(__always)
+internal func _lexicographicalCompare(
+  _ leftHS: Array<UInt16>,
+  _ rightHS: Array<UInt16>
+) -> _Ordering {
+  return leftHS.withUnsafeBufferPointer { leftPtr in
+    return rightHS.withUnsafeBufferPointer { rightPtr in
+      return _lexicographicalCompare(leftPtr, rightPtr)
+    }
+  }
+}
+
+internal func _parseRawScalar(
+  _ buf: UnsafeMutablePointer<_Normalization._SegmentOutputBuffer>,
+  startingFrom idx: Int = 0
+) -> (UnicodeScalar, scalarEndIndex: Int) {
+  return _parseRawScalar(_castOutputBuffer(buf), startingFrom: idx)
+}
+internal func _parseRawScalar(
+  _ buf: UnsafeBufferPointer<UInt16>,
+  startingFrom idx: Int = 0
+) -> (UnicodeScalar, scalarEndIndex: Int) {
+  let ptr = buf.baseAddress._unsafelyUnwrappedUnchecked
+  _sanityCheck(idx >= 0 && idx < buf.count, "out of bounds index")
+  let cu: UInt16 = ptr[idx]
+  if _slowPath(idx+1 == buf.count) {
+    return (UnicodeScalar(_unchecked: UInt32(cu)), idx+1)
+  }
+  guard _isLeadingSurrogate(cu) else {
+    return (UnicodeScalar(_unchecked: UInt32(cu)), idx+1)
+  }
+  let nextCu: UInt16 = ptr[idx+1]
+  guard _isTrailingSurrogate(nextCu) else {
+    // Invalid surrogate pair: just return the invalid value
+    return (UnicodeScalar(_unchecked: UInt32(cu)), idx+1)
+  }
+
+  // print("""
+  //   leading: \(hex(cu)), trailing: \(hex(nextCu))
+  // """)
+
+  // Decode
+  let value: UInt32 = _decodeSurrogatePair(leading: cu, trailing: nextCu)
+  _sanityCheck(Int32(exactly: value) != nil, "top bit shouldn't be set")
+  return (UnicodeScalar(_unchecked: value), idx+2)
+}
+internal func _parseRawScalar(
+  _ guts: _StringGuts,
+  startingFrom idx: Int = 0
+) -> (UnicodeScalar, scalarEndIndex: Int) {
+  var buffer = _FixedArray2<UInt16>(allZeros:())
+  buffer[0] = guts[idx]
+  buffer[1] = guts[idx+1]
```

Comment:
This is a bad idea and _StringGuts should remove this subscript. We should branch once at the top on the representation, and not repeatedly inside our tight loops. Why is this needed?